src/databricks/sql/thrift_backend.py

```diff
@@ -279,24 +287,41 @@ def attempt_request(attempt):
             # - non-None method_return -> success, return and be done
             # - non-None retry_delay -> sleep delay before retry
             # - error, error_message always set when available
+
+            error = None
+
+            # If retry_delay is None the request is treated as non-retryable
+            retry_delay = None
             try:
                 logger.debug("Sending request: {}".format(request))
                 response = method(request)
                 logger.debug("Received response: {}".format(response))
                 return response
-            except Exception as error:
+            except socket.timeout as err:
+                # socket.timeout means the connection was timed out by the socket package
+                retry_delay = self._retry_delay_default
+                error_message = str(err)
+                error = err
+            except OSError as err:
```

Comment:
As this is Operating system error, is the semantic of the error (code, message, etc) the same on different OSs (linux, win)?

Reply:
This is not guaranteed for _all_ operating systems, but for UNIX and Windows the codes are the same. See [Microsoft's Errno documentation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/errno-constants?view=msvc-170)

> Some of the values in ERRNO.H are present to maintain compatibility with the UNIX family of operating systems. The errno values in a 32-bit Windows operating system are a subset of the values for errno in UNIX systems.

Therefore `Errno 110` is the correct `EHOSTUNREACH` code for any UNIX system + Windows

Reply:
After we add HTTP proxy support I would like to establish multi-platform e2e tests which effectively duplicate an `EHOSTUNREACH` code without any mocks. Until then, it is possible that platform-specific formatting of socket timeout exceptions could be missed by this Exception handler.

In such case, the default is to not retry.

Reply:
Other question: in what cases do we _not_ want to retry this. E.g. can it also happen when a user specifies an incorrect host? In that case we probably don't want to retry (/hang) for 5*30 = 150s ?

Reply:
Good points @sander-goos 

> Besides the fact that the errorcode will map to the same error, are we also sure that the string rendering of the error is the same? (As that is what is being checked with if "Errno 110" in str(err):)

After some rework, the code now uses Python's built-in `errno` to encapsulate the error code check. We check for `errno.ETIMEDOUT` which will evaluate to `110` on Linux, `65` on MacOS, and `10060` on Windows hosts.

> Other question: in what cases do we not want to retry this. E.g. can it also happen when a user specifies an incorrect host? In that case we probably don't want to retry (/hang) for 5*30 = 150s ?

The current implementation is not picky about which `OSError`s it retries. But it will log a warning if an unexpected code is found. In the precise situation you outline, only a `GetOperationStatus` would be retried. Any other request, such as opening a connection / session or creating a cursor would re-raise the exception back to the caller immediately.